File: src/csharp/core/src/Avatar.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;
using DiceBear.Internal;

namespace DiceBear
{
    /// <summary>
    /// Top-level entry point for rendering an avatar from a style and options.
    /// </summary>
    /// <remarks>
    /// Construction immediately resolves and renders the SVG; the accessors
    /// return different serializations of that result.
    /// </remarks>
    public sealed class Avatar
    {
        private readonly string _svg;
        private readonly OrderedMap<object?> _resolvedOptions;

        /// <summary>
        /// Validates the options, then resolves and renders an avatar for the
        /// given style. Passing no options is the same as passing an empty
        /// object.
        /// </summary>
        /// <exception cref="OptionsValidationException">
        /// The options violate the options schema.
        /// </exception>
        /// <exception cref="CircularColorReferenceException">
        /// A color in the style definition references itself.
        /// </exception>
        /// <exception cref="CircularComponentReferenceException">
        /// A component in the style definition references itself.
        /// </exception>
        public Avatar(Style style, JsonObject? options = null)
        {
            if (style is null)
            {
                throw new ArgumentNullException(nameof(style));
            }

            var resolver = new Resolver(style, new Options(options));

            _svg = new Renderer(style, resolver).Render();
            _resolvedOptions = resolver.Resolved();
        }

        /// <summary>
        /// Renders an avatar from options given as JSON text.
        /// </summary>
        /// <exception cref="OptionsValidationException">
        /// The text is not valid JSON, or the options violate the schema.
        /// </exception>
        public static Avatar FromJson(Style style, string optionsJson)
        {
            JsonNode? node;

            try
            {
                node = JsonNode.Parse(optionsJson);
            }
            catch (JsonException exception)
            {
                throw new OptionsValidationException(new[]
                {
                    new ValidationErrorDetail(message: exception.Message),
                });
            }

            if (node is not JsonObject options)
            {
                // The constructor reads a null argument as "no options", which
                // is what its default parameter means. JSON text always carries
                // a value, so `null`, `[]` and `5` are inputs the schema
                // rejects rather than an absent argument.
                OptionsValidator.Validate(node);

                options = new JsonObject();
            }

            return new Avatar(style, options);
        }

        /// <summary>
        /// Returns the rendered SVG markup.
        /// </summary>
        public string ToSvg() => _svg;

        /// <summary>
        /// Returns the rendered SVG markup.
        /// </summary>
        public override string ToString() => _svg;

        /// <summary>
        /// Returns the SVG encoded as a <c>data:image/svg+xml</c> URI.
        /// </summary>
        public string ToDataUri() => "data:image/svg+xml;charset=utf-8," + Num.EncodeUriComponent(_svg);

        /// <summary>
        /// Returns the fully resolved options used to render the avatar, in
        /// resolution order. The raw seed is deliberately excluded.
        /// </summary>
        /// <remarks>
        /// A fresh object on every call, so a caller mutating the result cannot
        /// corrupt the avatar's internal state.
        /// </remarks>
        public JsonObject ResolvedOptions()
        {
            var result = new JsonObject();

            foreach (var entry in _resolvedOptions)
            {
                var value = ToJsonNode(entry.Value);

                // An option the user left unset is stored as null and dropped
                // here, the way JSON.stringify drops an undefined value in the
                // reference.
                if (value is not null)
                {
                    result[entry.Key] = value;
                }
            }

            return result;
        }

        /// <summary>
        /// Returns <c>{ "svg", "options" }</c> — the SVG and the resolved
        /// options used to render it — as JSON text.
        /// </summary>
        public string ToJson()
        {
            var envelope = new JsonObject
            {
                ["svg"] = JsonValue.Create(_svg),
                ["options"] = ResolvedOptions(),
            };

            return JsonWrite.Node(envelope);
        }

        private static JsonNode? ToJsonNode(object? value)
        {
            switch (value)
            {
                case null:
                    return null;
                case string text:
                    return JsonValue.Create(text);
                case bool flag:
                    return JsonValue.Create(flag);
                case double number:
                    return JsonValue.Create(number);
                case IReadOnlyList<string> list:
                    {
                        var array = new JsonArray();

                        foreach (var item in list)
                        {
                            array.Add(JsonValue.Create(item));
                        }

                        return array;
                    }

                default:
                    return JsonValue.Create(value.ToString());
            }
        }
    }
}

File: src/js/core/src/Avatar.ts
import { Style } from './Style.js';
import { Options } from './Options.js';
import { Resolver } from './Resolver.js';
import { Renderer } from './Renderer.js';
import type { StyleOptions } from './StyleOptions.js';

interface AvatarJson<D = unknown> {
  readonly svg: string;
  readonly options: StyleOptions<D>;
}

/**
 * Top-level entry point for rendering an avatar from a style and options.
 *
 * Construction immediately resolves and renders the SVG; the various
 * accessor methods return different serializations of that result.
 */
export class Avatar<D = unknown> {
  #svg: string;
  #resolvedOptions: StyleOptions<D>;

  /**
   * Pass a {@link Style} instance and reuse it across avatars. The parsed
   * style carries the validation, so nothing is re-checked per render.
   */
  constructor(style: Style<D>, optionsInput?: StyleOptions<D>) {
    if (!(style instanceof Style)) {
      throw new TypeError(
        '[DiceBear] `new Avatar()` expects a Style. Wrap the definition first: ' +
          '`new Avatar(new Style(definition), options)`.',
      );
    }

    const options = new Options<D>(optionsInput);
    const resolver = new Resolver(style, options);

    this.#svg = new Renderer(style, resolver).render();
    this.#resolvedOptions = resolver.resolved();
  }

  /**
   * Returns the rendered SVG markup.
   */
  toString(): string {
    return this.#svg;
  }

  /**
   * Returns the avatar as a JSON-serializable object containing the SVG and
   * the fully resolved options used to render it.
   */
  toJSON(): AvatarJson<D> {
    return {
      svg: this.#svg,
      options: structuredClone(this.#resolvedOptions),
    };
  }

  /**
   * Returns the SVG encoded as a `data:image/svg+xml` URI.
   */
  toDataUri(): string {
    return `data:image/svg+xml;charset=utf-8,${encodeURIComponent(this.#svg)}`;
  }
}
